### ISSUE TITLE
resource "aiven_service" don't waiting for the complete creation (or deletion) of the service

### ISSUE THREAD

**Issue by benesis002** [aiven/terraform-provider-aiven#152]
When Terraform create a "aiven_service" resource, the provider set the task as finish before waiting the complete service creation on Aiven side..
It's not a good practice because when Terraform want to run a new Task using the new aiven service created (Grafana for exemple), the service is not ready and the new task fail.

Can you fix it please ?

**Comment by fingon** [comment]
Is there something in particular in which the terraform provider does wrong? I haven't looked at the code, but at least for some service types there are problems also in the REST API endpoint that are used by both this and the aiven client ( https://github.com/aiven/aiven-client ).

**Comment by ivan-savciuc** [comment]
Hi,

Aiven Terraform provider relies on the Aiven go client https://github.com/aiven/aiven-client and consequently on Aiven API itself. Provider determines whether service is created based on the API response field `state` and when it is equal to "RUNNING" we change the resource status to created. Can you please give an example of a situation when a Grafana service is not entirely created?

The generic test doesn't give any negative results:
```
resource "aiven_service" "bar" {
  project = aiven_project.es-project.project
  cloud_name = "google-europe-west1"
  plan = "startup-1"
  service_name = "test-acc-sr-1"
  service_type = "grafana"
}


data "aiven_service" "grafana" {
  project = aiven_project.es-project.project
  service_name = aiven_service.bar.service_name
}

output "grafana-uri" {
  value = data.aiven_service.grafana.service_uri
}
```

Where `grafana-uri` is immediately actionable after appearing in the output.

**Comment by fingon** [comment]
There is bit of a race condition, as the server-side seems to e.g. enable other services' data sources only after entering RUNNING. We should fix this though, as it is also sanity check of sorts for Grafana actually being functional.

**Comment by benesis002** [comment]
Hello, the problem occurs when i run this test : 
`
resource "aiven_service" "grafana" {
  project           = local.aiven_project_name
  project_vpc_id    = var.aiven_vpc_id
  cloud_name        = local.aiven_cloud_name
  plan              = "startup-4"
  service_name      = local.grafana_aiven_service_name
  service_type      = "grafana"
  grafana_user_config {
    ip_filter       = ["0.0.0.0/0"]
    public_access {
      grafana = true
    }
  }
}

provider "grafana" {
  url  = "https://${local.aiven_grafana_public_url}"
  auth = "${aiven_service.grafana.service_username}:${aiven_service.grafana.service_password}"
}

resource "grafana_dashboard" "cert_manager" {
  config_json = templatefile("${path.module}/grafana_dashboard/cert-manager.json", { DS_PROMETHEUS = "Prometheus" })
  depends_on = [aiven_service.grafana, null_resource.wait_grafana_public_url]
}
`
We get an 400 error for the first Terraform apply.
The workaround is add this null resource :

`# Deploy dashboards
resource "null_resource" "wait_grafana_public_url" {
  provisioner "local-exec" {
    command = "${path.module}/scripts/wait_grafana_public_url.sh ${local.aiven_grafana_public_url}"
  }
}
`

**Comment by ivan-savciuc** [comment]
Hello, 

Thank you for providing details for this issue, now it makes more sense. 

There is a difference between service status and the status of configuration options for a service that been applied. For example, when a user creates a Grafana service, it will have state different from "RUNNING" during VM provisioning, Grafana service installation, etc. Until Garana service will be running and can handle user requests. And then when service is running, we start applying service configuration options like:

```
public_access {
     grafana = true
}
```

And the state of applied configuration options is not reflected anywhere in `aiven_service` resource, that is why credentials and URL that you are giving to Grafana provider may not be immediately available. 

This is a limitation of Aiven terraform provider that we cannot address immediately and will require a general solution for a subset of problems related to service configuration options. Any ideas are welcome.

**Comment by benesis002** [comment]
Hello,

We have the same problem during the destroy action.

From the same terraform stack, we create a VPC + VPC peering on Aiven side.

```
resource "aiven_project_vpc" "main" {
  project      = local.aiven_project_name
  cloud_name   = "google-${var.gcp_project.default_region}"
  network_cidr = local.aiven_vpc_network_cidr
}

resource "aiven_vpc_peering_connection" "main" {
  vpc_id             = aiven_project_vpc.main.id
  peer_cloud_account = var.gcp_project.id
  peer_vpc           = module.vpc_creation.network_name
}

```

Error is :
`Error: 409: {"errors":[{"message":"VPC cannot be deleted while there are services migrating from it","status":409}],"message":"VPC cannot be deleted while there are services migrating from it"} - `

A workaround is add a null resource too, and aiven_project_vpc.main depend of this null resource

Benjamin

**Comment by benesis002** [comment]
Hello,

Maybe we found the root cause.
For our problem during the deletion phase, we have try to create a "null_resource" which call Aiven API to check if any services are always present in the Aiven project.

We have detected that the API, like the Aiven Web UI console, like the Aiven provider don't waiting for the complete deletion on the Aiven/GCP side, before to respond that the deletion is completed.

In this case, when from the API or the Web console you don't see any services, it's not true, somes GCP ressources are always present on your side. The aiven deletion process is not totally finished.
It is why we cannot delete an Aiven VPC just after deleted all Aiven services.

So it's not a problem from your Terraform provider, it's a problem from your Aiven API or internal process. Can you create an issue in your team for this ?

Benjamin

**Comment by ivan-savciuc** [comment]
Hello,

Indeed, Aiven API is eventually consistent like any cloud API nowadays (AWS, GCP, etc.), and if it is possible to determine based on API Response what is the state of an entity like service, VPC, etc. Then Terraform Provider should provide a smooth experience of managing infrastructure without any problems like you've encountered. We can fix VPC deletion because we do  have `Error: 409` in response and can handle in Terraform Provider code but in the case of 
```
public_access {
      grafana = true
    }
```
we do not have any clear indication from API side when this configuration setting has been applied, therefore cannot be solely fixed on Terraform Provider side and will require the transmission of a different status from API side.

I will create corresponding tickets in our system but cannot provide a link since we do not have a public issue tracker.

Ivan

**Comment by ivan-savciuc** [comment]
In the last Provider release v1.2.2 https://github.com/aiven/terraform-provider-aiven/releases/tag/v1.2.2, we have added some features which will solve some of the problems from this ticket. 

In the new version: 
- Grafana service now waits until Grafana is reachable publicly (only works in case public_access.grafana
configuration options are set to true, and IP filter is set to default 0.0.0.0/0) during resource creation or update.
- Added Project VPC resource graceful deletion.
